### DISCUSSION TITLE
The last column in a workspace which is non-full-width can have empty space on the right

### DISCUSSION THREAD

**Discussion by salman-farooq-sh** [YaLTeR/niri#466] [Ideas]
...when using the touchpad gestures. Keyboard shorcuts work "correctly" (correct being what I want :), hehe).

I guess this was done to indicate that new windows spawn on the empty space at right, similar to the empty workspace at bottom.

But:
1. Having consistency between keyboard and touchpad behaviours might be nicer.
2. In the bottom/empty workspace case, we kinda need it with the current UX. In this case we don't necessarily need it per se.
3. You need to be careful and precise with sideways touchpad gestures if you are working with a non-full-width column at the extreme start or end of a workspace, otherwise it throws the neighboring column out of the view hindering multitasking.

I don't have an ultra-wide but I guess this might annoy those users too as they would have non-full-width-columns more often.

A solution:
Only switch one (or N) column/s at most no matter the "strength" of the touchpad gesture, or expose some granular gesture config parameters to the same effect. I don't know which I would prefer.

**Comment by YaLTeR** [comment]
I thought about this when implementing the gesture and it wasn't entirely clear to me that restricting the snapping like that is always desirable, so I ended up not doing it (which is also simpler code-wise). Sometimes I do want to have the last window on the left because I expect to open dialogs from it, in which case it is less annoying when it is stationary. It's also possible to achieve with a keyboard:
* Pressing Mod+F twice, or
* Opening a new window, switching focus back and forth to reset the "activate the last window and restore the camera" logic, then closing that new window.

**Reply by salman-farooq-sh** [reply]
I think having the proper solution of being able to have only a single column switch at most with a single gesture (either with high-level config or granular gesture control) will solve for me not only this issue but another one too (which I guess is the real/parent issue) that gestures need to be precise in general in niri. I can totally see the current behaviour being a feature too for some people.

I can understand what you mean by code-simplicity, because I am a web developer and we had to implement similar "snapping" on a web page at my job, we were not able to achieve something super nice in short time. Maybe we can look at GNOME's implementation? GNOME's is very nice, only behind macOS I would say. I also have nostalgia of my brother's HTC One M7 having very nice snapping in the Blinkfeed in Sense UI circa android 5.0 days.

**Comment by salman-farooq-sh** [comment]
Instead of having one column switch at a time with gestures (which is not un-welcome), upon further thought and daily-driving-while-keeping-this-in-mind I now think it will be really helpful to have at least config for trying to not have empty space on left or right of windows whenever possible.

This will make it the same as the default config's keyboard behaviour, Mod+WheelScroll behaviour, and the latest release's Mod+middle-mouse-button-drag behaviour.

**Reply by YaLTeR** [reply]
I did have another request for an options like this, which would also always shift the windows to the rightmost position upon cases like I described above, so maybe, yeah.

> and the latest release's Mod+middle-mouse-button-drag behaviour

This works the same as the touchpad gesture fwiw.

**Reply by salman-farooq-sh** [reply]
Yes, you are right about Mod+middle-mouse-button-drag, my bad. My mouse's battery was low and the cursor being janky due this somehow made it seem like that.

**Reply by salman-farooq-sh** [reply]
> the same as the default config's keyboard behaviour, (and) Mod+WheelScroll behaviour

still though

**Comment by Schweber** [comment]
I notice this behaviour when i have the right most window of a workspace maximized and then un-maximize it. I then get empty space to the right of it and have to press "focuse left; focus right" to make it go away and still have the window i  was working with selected.

I'd much rather have the screen "scroll to the left" automatically so that the whole screen is occupied by windows and i don't have to do these two key presses every time.

**Comment by YaLTeR** [comment]
I just pushed 952916fd1c61ebd61e7b40211d97fdd0c15e85d1 which prevents this empty space when using the gesture (touchpad horizontal swipe, or Mod + middle mouse + drag). Please try it and tell me how it feels.

It is still possible to leave empty space using maximize-unmaximize, which is still intended.

**Reply by salman-farooq-sh** [reply]
Works great with the touchpad gesture, sweet. Thanks.

Though I am also in favor of no empty space on right when unmaximizing as well. Let's think about how other vertically scrollable views work in various apps for UX inspiration. Take the android notifications dropdown for example, when there are a bunch of notifications such that the view is scrollable, the notifications on the top slide down to fill the empty space when the last notification in the list is dismissed. This also applies to the right most window getting closed.

To me, it also just makes sense anyways to display _something_ rather than nothing/empty space.

**Reply by cyuria** [reply]
I disagree with the notion that you should always display _something_, but at the same time I think ir might feel a bit more consistent if unmaximizing snapped to the right side as well. If that is done though, I think the default unmaximizing (and also possibly general resize) should expand from the centre instead of from the left, while preserving alignment so that size increases never overflow off screen. I'm not entirely sure how this would play out for shrinking a window though.

**Reply by Schweber** [reply]
> It is still possible to leave empty space using maximize-unmaximize, which is still intended.

I don't see the benefit in that. I'd rather have my screen always utilized to the fullest, that's why i'm using a tiling window manager after all.

**Reply by YaLTeR** [reply]
Sometimes I know that I'll be spawning temporary windows on the right, so I'll leave some space there to reduce the scrolling when they open/close.

**Reply by Schweber** [reply]
Could you please make an option that toggles this behaviour?

**Reply by salman-farooq-sh** [reply]
I agree, both workflows are valid and this should ideally be a config option in some capacity.
